### PR TITLE
Backport #5384 - Support of generic case class for the JSON macros

### PR DESCRIPTION
**Pull request by cchantep**

### CODE REVIEW

File: framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala [hunk 1]
```diff
@@ -105,12 +105,15 @@ object JsMacroImpl {
     val applies =
       companionType.decl(TermName("apply")) match {
         case NoSymbol => c.abort(c.enclosingPosition, "No apply function found")
-        case s => s.asTerm.alternatives
+        case s => s.asTerm.alternatives.filter {
+          _.asMethod.paramLists.size == 1
```
Comment (OlegYch) [inline framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala]: why the restriction?

**Comment by alexdupre** [comment]
This change broke serialization of case classes with multiple arguments list, is it intended?

**Comment by cchantep** [comment]
I don't see how it can break the JSON serialization, as such classes were just not supported before (compilation error), while the API is unchanged.

**Comment by OlegYch** [comment]
it was working before http://scastie.org/21867
doesn't work now http://scastie.org/21868 :(

**Comment by cchantep** [comment]
Writes could have "accidentally" worked with such multi parameters constructors, even if never documented, but the `reads` and `format` didn't and cannot. The macros now enforce explicitly and in all the case this same restriction: no multi param constructor for case classes (break coherence between `apply`/`unapply`.

**Comment by OlegYch** [comment]
i'm just pointing out that 2.5.5 is no longer compatible with 2.5.4

**Comment by cchantep** [comment]
It's binary and API compatible, and also fixes macro incoherencies about an undocumented/not recommanded case. Not it can be said to "break", but it can be documented now.

**Comment by OlegYch** [comment]
i've created an issue #6476 as it is very clearly a breaking change for us

**Comment by kpmeen** [comment]
I understand that the restrictions described previously in this thread applies for more than one parameter lists. But I'm also getting the  'No apply function found matching unapply parameters' message when I have a generic case class more than 1 generic parameters in the same param list. Like so:

```scala
case class Interval[T](base: T, tolerance: Option[T])

implicit val intFmt: Format[Interval[Int]] = Json.format[Interval[Int]]
```

I would expect that to work, or?

**Comment by cchantep** [comment]
@kpmeen I guess this is more related to #5261: the following case class with multiple generic parameter (not recursive) is working fine. There is no change about that for now.

```scala
case class Interval[T](base: T, tolerance: T)
```

**Comment by kpmeen** [comment]
ah...ok. Thanks @cchantep.

**Pull request merged**